src/Aspire.Hosting/Dashboard/KubernetesDataSource.cs

```diff
@@ -0,0 +1,581 @@
+
+                    try
+                    {
+                        await handler(eventType, resource).ConfigureAwait(false);
+                    }
+                    finally
+                    {
+                        semaphore.Release();
+                    }
+                }
+            }
+            catch (Exception ex) when (ex is not OperationCanceledException)
+            {
+                _logger.LogError(ex, "Watch task over kubernetes resource of type: {resourceType} terminated", typeof(T).Name);
+            }
+        }
+    }
+
+    private async Task ProcessContainerChange(WatchEventType watchEventType, Container container)
+    {
+        if (!ProcessResourceChange(_containersMap, watchEventType, container))
+        {
+            return;
+        }
+
+        UpdateAssociatedServicesMap(ResourceKind.Container, watchEventType, container);
+
+        var objectChangeType = ToObjectChangeType(watchEventType);
+        var containerViewModel = ConvertToContainerViewModel(container);
+
+        await _onResourceChanged(containerViewModel, objectChangeType).ConfigureAwait(false);
+    }
+
+    private async Task ProcessExecutableChange(WatchEventType watchEventType, Executable executable)
+    {
+        if (executable.IsCSharpProject())
+        {
+            await ProcessProjectChange(watchEventType, executable).ConfigureAwait(false);
+            return;
+        }
+
+        if (!ProcessResourceChange(_executablesMap, watchEventType, executable))
+        {
+            return;
+        }
+
+        UpdateAssociatedServicesMap(ResourceKind.Executable, watchEventType, executable);
+
+        var objectChangeType = ToObjectChangeType(watchEventType);
+        var executableViewModel = ConvertToExecutableViewModel(executable);
+
+        await _onResourceChanged(executableViewModel, objectChangeType).ConfigureAwait(false);
+    }
+
+    private async Task ProcessProjectChange(WatchEventType watchEventType, Executable executable)
+    {
+        if (!ProcessResourceChange(_executablesMap, watchEventType, executable))
+        {
+            return;
+        }
+
+        UpdateAssociatedServicesMap(ResourceKind.Executable, watchEventType, executable);
+
+        var objectChangeType = ToObjectChangeType(watchEventType);
+        var projectViewModel = ConvertToProjectViewModel(executable);
+
+        await _onResourceChanged(projectViewModel, objectChangeType).ConfigureAwait(false);
+    }
+
+    private async Task ProcessEndpointChange(WatchEventType watchEventType, Endpoint endpoint)
+    {
+        if (!ProcessResourceChange(_endpointsMap, watchEventType, endpoint))
+        {
+            return;
+        }
+
+        if (endpoint.Metadata.OwnerReferences is null)
+        {
+            return;
+        }
+
+        foreach (var ownerReference in endpoint.Metadata.OwnerReferences)
+        {
+            // Find better way for this string switch
+            switch (ownerReference.Kind)
```

Comment:
DCP model has definitions for well-known kinds: https://github.com/dotnet/aspire/blob/main/src/Aspire.Hosting/Dcp/Model/GroupVersion.cs#L24

Reply:
Thanks. They cannot be used in `switch` though as they're not compile-time constants. If you're happy for me to change them from get-only properties to consts, I can use them here. Or we fall back to if/else chaining.